src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/DataContentAsyncEnumerableStream.cs

```diff
@@ -0,0 +1,228 @@
+#else
+internal sealed class DataContentAsyncEnumerableStream<T> : Stream
+#endif
+    where T : DataContent
+{
+    private readonly IAsyncEnumerator<T> _enumerator;
+    private bool _isCompleted;
+    private byte[] _remainingData;
+    private int _remainingDataOffset;
+    private long _position;
+    private T? _firstDataContent;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DataContentAsyncEnumerableStream{T}"/> class, where T is <see cref="DataContent"/>.
+    /// </summary>
+    /// <param name="dataAsyncEnumerable">The async enumerable to stream.</param>
+    /// <param name="firstDataContent">The first chunk of data to reconsider when reading.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <remarks>
+    /// <paramref name="firstDataContent"/> needs to be considered back in the stream if <paramref name="dataAsyncEnumerable"/> was iterated before creating the stream.
+    /// This can happen to check if the first enumerable item contains data or is just a reference only content.
+    /// </remarks>
+    internal DataContentAsyncEnumerableStream(IAsyncEnumerable<T> dataAsyncEnumerable, T? firstDataContent = null, CancellationToken cancellationToken = default)
+    {
+        _enumerator = Throw.IfNull(dataAsyncEnumerable).GetAsyncEnumerator(cancellationToken);
+        _remainingData = Array.Empty<byte>();
+        _remainingDataOffset = 0;
+        _position = 0;
+        _firstDataContent = firstDataContent;
+    }
+
+    /// <inheritdoc/>
+    public override bool CanRead => true;
+
+    /// <inheritdoc/>
+    public override bool CanSeek => false;
+
+    /// <inheritdoc/>
+    public override bool CanWrite => false;
+
+    /// <inheritdoc/>
+    public override long Length => throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override long Position
+    {
+        get => _position;
+        set => throw new NotSupportedException();
+    }
+
+    /// <inheritdoc/>
+    public override void Flush() => throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override long Seek(long offset, SeekOrigin origin) =>
+        throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override void SetLength(long value) =>
+        throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException("Use ReadAsync instead for asynchronous reading.");
+    }
+
+    /// <inheritdoc/>
+#if NET8_0_OR_GREATER
+    public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        => ReadAsync(buffer, cancellationToken).AsTask();
+#else
+    public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```

Comment:
This should be able to use the Memory implementation on netstandard/netfx as well, no? That method would be a public override on core, and a private helper downlevel.